Spare IrisAO segment numbers testing

ivalaginja opened the pull request:
Upstream for https://github.com/spacetelescope/hicat-package/pull/340

These are changes that will provide us with correctly controllable hardware, which we can then use as a baseline to do any refactoring in the segmented DM controls.

RemiSoummer committed:
active segments per pupil fix for HiCAT spare

ivalaginja committed:
unrotate hex DM in SegmentedAperture()

ivalaginja committed:
add units to docstring of update_one_segment()

ivalaginja committed:
take out the rounding of WFE display numbers

ivalaginja commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -90,7 +90,7 @@ def create_aperture(self):
                                                           gap=self.gap,
                                                           segmentlist=self._segment_list,
-                                                          rotation=-90)
+                                                          rotation=0)
```
@kjbrooks do we know the "real world" orientation of the JOST IrisAO? Does JOST have a single corner segment in the upwards direction like the spare HiCAT IrisAO here, or one of the flat sides with segments in a straight line? I am not sure what kind of rotations the camera, Sharpcap... introduce, so we might need to check this in the lab right in front of the camera next time somebody is in. We can probably do this bye applying a global defocus on the IrisAO (either by commanding the mirror or by translating the camera).
In any case, we might want to consider to pull this parameter out so that it is changeable. Also, it's only used for the command display, so this is not crucial for the controls, but it's easier to work with the commands if they're oriented correctly.

kjbrooks replied:
For JOST, I think it's flat side up. I did a lot of testing and -90 is correct for JOST so I recommend that you make this a variable instead of making a hard coded change since that will mess up JOST. I am all for looking more into this, but if we want it to work on both of them right now, we will need to make this a variable.

ivalaginja committed:
pass full path to custom flat through to controls

RemiSoummer committed:
add comment about units

ivalaginja committed:
specify details of rounded list

ivalaginja committed:
add comments about *custom* flat map, and units

kjbrooks reviewed (changes requested):
In order to make this compatible with JOST and HiCAT, I recommend changing the rotation to a parameter that is passed in by the user.

ivalaginja committed:
make rotation an input parameter

ivalaginja commented:
@kjbrooks I made it an input parameter now

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -306,14 +315,15 @@ def to_command(self):
-        to or will replace the current PTT tuple. To identify the segment to be changed, give
-        the *index* of that segment in the active segment list. This will be added
+        to or will replace the current PTT tuple on this segment only. To identify the segment to be changed, give
+        the *command index* (not its name like in the GUI) of that segment in the active segment list. This will be added
```
typo. change to "it's"

ivalaginja replied:
Are sure? This is not intended to be "it is", just "the/its name".

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -452,7 +464,7 @@ def plot_psf(self, rotation_angle, figure_name_prefix, out_dir, vmin=10e-8,
 
 def load_command(segment_values, dm_config_id, wavelength, testbed_config_id,
-                 apply_flat_map=True):
+                 apply_flat_map=True, filename_flat=''):
```
Since `filename_flat` is only used when `apply_flat_map = True`, I recommend changing the default to `None`

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -466,12 +478,14 @@ def load_command(segment_values, dm_config_id, wavelength, testbed_config_id,
                          regarding the segmented DM can be found.
-    :param apply_flat_map: Apply a flat map in addition to the data.
+    :param apply_flat_map: bool, whether to apply a flat map in addition to the data when sending command to hardware
+    :param filename_flat: string, full path to the custom flat map
```
add that this is only needed if `apply_flat_map=True`

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -197,42 +203,45 @@ class SegmentedDmCommand(SegmentedAperture):
 
-    :param apply_flat_map: If true, add flat map correction to the data before creating command
+    :param apply_flat_map: If true, add the custom flat map correction to the data before creating the command
+    :param filename_flat: string, full path to custom flat map
```
same a below, mention that this is only used if `apply_flat_map`

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -197,42 +203,45 @@ class SegmentedDmCommand(SegmentedAperture):
-    :attribute filename_flat: str, path to flat
-    :attribute total_number_segments: int, total number of segments in DM
+    :attribute command: dict, final command with flat if apply_flat_map = True; units are determined with the attribute dm_command_units
+    :attribute filename_flat: str, full path to custom flat
```
same commend as other places, only used if `apply_flat_map=True`

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -197,42 +203,45 @@ class SegmentedDmCommand(SegmentedAperture):
     """
 
-    def __init__(self, dm_config_id, wavelength, testbed_config_id, apply_flat_map=False):
+    def __init__(self, dm_config_id, wavelength, testbed_config_id, apply_flat_map=False, filename_flat='', rotation=0):
```
set `filename_flat` to `None`

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -157,7 +163,7 @@ def get_segment_list(self):
 
         :param num_rings: int, The number of rings in your pupil
-        :return: list, the list of segments
+        :return: list, the list of segments # TODO: the list of segment *names* (as e.g. in the GUI)?
```
This is not segment names. This is the total number of segments in a pupil given the number of rings. Remove #TODO. Clarify parameters description as you see necessary

kjbrooks commented on catkit/hardware/iris_ao/segmented_dm_command.py:
```diff
@@ -9,7 +9,8 @@
                                                      testbed_config_id,
-                                                     apply_flat_map=True)
+                                                     apply_flat_map=True,
+                                                     filename_flat=FLAT_COMMAND_FILE)
```
Since you don't define `FLAT_COMMAND_FILE` here I recommend you just create a dummy string to make it clear what you mean.

kjbrooks reviewed (approved):
I found a few other things I didn't see the first time. Sorry about that, I am not sure how I missed them the first time. I will approve this PR but I am hoping you will make the changes I requested.

ivalaginja commented:
@kjbrooks I addressed all your comments but opened a new PR for them so that I can merge this one without having to wait for another rounds of tests and re-approval. Let me know if you want to see anything else over there:
https://github.com/spacetelescope/catkit/pull/140
